File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/controller/TokenRestController.java
package com.raxim.myscoutee.profile.controller;

import java.util.UUID;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.raxim.myscoutee.common.config.firebase.dto.FirebasePrincipal;
import com.raxim.myscoutee.common.util.JsonUtil;
import com.raxim.myscoutee.profile.data.document.mongo.Token;
import com.raxim.myscoutee.profile.data.document.mongo.User;
import com.raxim.myscoutee.profile.repository.mongo.TokenRepository;
import com.raxim.myscoutee.profile.util.AppConstants;

@RestController
@RequestMapping("tokens")
public class TokenRestController {

    private final TokenRepository tokenRepository;
    private final ObjectMapper objectMapper;

    public TokenRestController(TokenRepository tokenRepository, ObjectMapper objectMapper) {
        this.tokenRepository = tokenRepository;
        this.objectMapper = objectMapper;
    }

    @PostMapping("")
    public ResponseEntity<Token> register(@RequestBody Token token, Authentication auth) {
        Token tokenObj = tokenRepository.findByDeviceId(token.getDeviceKey());

        if (tokenObj == null) {
            FirebasePrincipal principal = (FirebasePrincipal) auth.getPrincipal();
            User user = principal.getUser();
            Token tokenToSave = JsonUtil.clone(token, objectMapper);
            tokenToSave.setId(UUID.randomUUID());
            tokenToSave.setUuid(user.getId());
            tokenToSave.setType(AppConstants.FIREBASE);
            tokenObj = tokenRepository.save(tokenToSave);
        }

        return new ResponseEntity<>(tokenObj, HttpStatus.CREATED);
    }
}

File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/controller/TokenRestController.kt
package com.raxim.myscoutee.profile.controller

import com.raxim.myscoutee.common.config.firebase.FirebasePrincipal
import com.raxim.myscoutee.profile.data.document.mongo.Token
import com.raxim.myscoutee.profile.repository.mongo.TokenRepository
import org.springframework.data.rest.webmvc.RepositoryRestController
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping

@RepositoryRestController
@RequestMapping("tokens")
class TokenRestController(val tokenRepository: TokenRepository) {

    @PostMapping("")
    fun register(@RequestBody token: Token, auth: Authentication): ResponseEntity<Token> {

        var tokenObj = this.tokenRepository.findByDeviceId(token.deviceKey!!)
        if(tokenObj == null) {
            val user = (auth.principal as FirebasePrincipal).user
            val tokenToSave = token.copy(uuid = user.id)
            tokenObj = this.tokenRepository.save(tokenToSave);
        }

        return ResponseEntity(tokenObj, HttpStatus.CREATED);
    }

}